### PR TITLE
chore: flatten diagnostics + error list when appending on an error list

### PR DESCRIPTION
**Pull request by katcipis**
So it is easier to build error list from operations that produce a diagnostic from HCL (which is also a list) or even our own list. It seems that the most desireable behavior is flattening, but I could be wong.

### CODE REVIEW

File: errors/error.go [hunk 1]
```diff
@@ -116,13 +119,23 @@ func E(args ...interface{}) error {
 		case hcl.Range:
 			e.FileRange = arg
 		case hcl.Diagnostics:
-			diags = arg
+			diags := arg
 			if diags.HasErrors() {
```
Comment (i4ki) [inline errors/error.go:123]: I just realized that `HasErrors()` looks for error severity, then if there are also Warnings just getting the diags[0] will get the wrong error. Not related to your change but have to be fixed.

File: errors/error.go [hunk 2]
```diff
@@ -96,13 +99,13 @@ const separator = ": "
 //
 // In order to avoid duplicated messages, we erase the fields present in the
 // underlying error if already set with same value in this error.
-func E(args ...interface{}) error {
+func E(args ...interface{}) *Error {
```
Comment (i4ki) [inline errors/error.go:102]: curious why this change is needed.

File: errors/list.go [hunk 3]
```diff
@@ -84,11 +92,30 @@ func (l *List) Detailed() string {
...
+//
+// Any error of type errors.List will be flattened inside
+// the error list.
 func (l *List) Append(err error) {
 	if err == nil {
 		return
 	}
-	l.errs = append(l.errs, err)
+	switch e := err.(type) {
+	case hcl.Diagnostics:
```
Comment (i4ki) [inline errors/list.go:106]: :thinking: 
strange this also here.
Why not use like `errs.Append(errors.E(diags))` ?

**Review by i4ki — APPROVED** [review]
LGTM but left some questions.
The issue with `HasErrors()` and accessing `diags[0]` can be fixed in a separate PR, but it's your call.

File: errors/error.go (hunk 2 above)
Comment (katcipis) [reply errors/error.go:102]: I was comparing with the result of List.Errors and wanted to build Error instances, I could also have used the type directly instead of errors.E(), but then the other question would be, why keep it as error ? When a type implements an interface it is rare to return the interface type itself on the constructor, you usually just return the concrete type and that will work out fine anyway when it is used as the interface. So I changed more because I don't see strong reasons for it to be error, it seems to enforce a constraint that is not necessary.

**Comment by katcipis** [comment]
I will merge this so we can see it in action on a PR Im already working on, then we can keep refining it and fix the HasErrors thing.
